### PR TITLE
chore: Generate less code (REST)

### PR DESCRIPTION
**Pull request by marikaner**
## Context

In the review of the previous generator pr, we discussed a more generic approach that should be implemented on a separated PR. This is it. I replaced the generated additional classed by generics and transformed the api request builder class to an object.

## Definition of Done

Please consider all items and remove only if not applicable.

- [x] Tests created/adjusted for your changes.
- [ ] Release notes updated.
  * Provide sufficient context so that each entry can be understood on its own.
  * Be specific about names of functions, classes, modules, etc.
  * Describe when or where this is relevant
  * Use indicative and present tense. For example, write "Provide function `name` that does X in order to Y" over "Now X can be done by calling a new function".
- [x] PR title adheres to [conventional commit guidelines](https://www.conventionalcommits.org) (please note that only `fix:` and `feat:` will end up in the release notes)
- [ ] If applicable: Properly documented (JSDoc of public API)
- [ ] If applicable: Check if `node run doc` still works.

### CODE REVIEW

File: packages/core/src/rest/rest-request-builder.ts [hunk 1]
```diff
@@ -1,9 +1,24 @@
-import { AxiosRequestConfig } from 'axios';
 import { Destination, DestinationNameAndJwt } from '../scp-cf';
 import { buildAxiosRequestConfig } from '../http-client';
 
-export abstract class RestRequestBuilder {
-  public customHeaders: Record<string, string> = {};
+type FunctionReturnType<ApiT, FnT extends keyof ApiT> = ApiT[FnT] extends (
```
Comment (FrankEssenberger) [inline packages/core/src/rest/rest-request-builder.ts:4]: Very impressive what is possible with the conditional type. Please add some comment what is done here:
basic open ApiT is the class e.g. `PriceCalculationApi` which has multiple functions e.g. `calculateViaRestWithTenant` represented by the two types `ApiT` and `FnT`. This ensures that on new RestRequestBuilder<PriceCalculationApi>(    PriceCalculationApi,
    'calculateViaRestWithTenant') you need to pass a key as second argument.

Comment (FrankEssenberger) [reply packages/core/src/rest/rest-request-builder.ts:4]: Request: Could you find or add a second method returning a union type e.g.
AxiosPromise<number|string>. I think the type magic should work but I would like to see a test.

File: packages/core/src/rest/api-request-builder.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import { RestRequestBuilder } from './rest-request-builder';
+
+export type ParametersType<
```
Comment (FrankEssenberger) [inline packages/core/src/rest/api-request-builder.ts:3]: Where are these types used. I do not see a reference

File: packages/core/src/rest/rest-request-builder.ts (hunk 1 above)
Comment (marikaner) [reply packages/core/src/rest/rest-request-builder.ts:4]: Yes, I will add some documentation around this. For the second part I will see what I can do. I guess I can only achieve this by providing a custom service spec file, which I would rather want to do on a separate PR (which can be merged before this one).

File: packages/core/src/rest/api-request-builder.ts (hunk 2 above)
Comment (marikaner) [reply packages/core/src/rest/api-request-builder.ts:3]: Yeah, at least the second one might not be necessary, I'll check.

File: packages/core/src/rest/rest-request-builder.ts (hunk 1 above)
Comment (FrankEssenberger) [reply packages/core/src/rest/rest-request-builder.ts:4]: Yes I think a custom spec file with all facets we want to check makes sense. A seperate PR for this makes also sense.

Comment (marikaner) [reply packages/core/src/rest/rest-request-builder.ts]: Hey @FrankEssenberger 

I found a couple of problems in the openapi generator (not our additions, but the java thing below). Let's discuss them in the refinement. I documented everything, but would like to refrain from adding tests on this PR, unless you have anything more realistic related to my changes in mind ;) I added a note about you union type idea to the already existing ticket on testing this.

File: packages/core/src/rest/api-request-builder.ts (hunk 2 above)
Comment (marikaner) [reply packages/core/src/rest/api-request-builder.ts]: I removed the second one and moved the structures a bit to separate better.

File: packages/rest-generator/src/request-builder/file.ts [hunk 3]
```diff
@@ -8,56 +7,24 @@ import { toPascalCase } from '@sap-cloud-sdk/core';
...
 import { coreImportDeclaration } from '../utils';
-import { apiRequestBuilderClass } from './api-request-builder-class';
-import { operationRequestBuilderClass } from './operation-request-builder-class';
+import { operationsVariable } from './operations';
 
-/**
- * Used by the generator for generating the request builder source file.
- * @param serviceMetadata The service metadata model converted from the open api file.
- * @returns source file structure of the request builder file
- */
```
Comment (FrankEssenberger) [inline packages/rest-generator/src/request-builder/file.ts:18]: why you removed the docs?

**Comment by FrankEssenberger** [comment]
Question: All the test are commented out and also the code generation is not present anymore. So it is hard to tell for me what the generator actually does without the result beeing checked in.

**Review by FrankEssenberger — APPROVED** [review]
LGTM as I understood the tests and sample service will come back and have been removed due to IP issues.

**Pull request merged**